sklearn_numba_dpex/common/kernels.py

```diff
@@ -223,43 +232,90 @@ def make_sum_reduction_2d_axis1_kernel(
+    a multiple of `sub_group_size`, such that `work_group_size // sub_group_size` is
+    a power of 2.
+
+    Depending on the size of the sum axis, it might be worth tuning `work_group_size`
+    for better performance. When `axis` is `1`, tuning `sub_group_size` with respect
+    to the size of the other axis might also be beneficial.
+
+    The algorithmic approach that is chosen to implement the underlying kernels takes
+    inspiration from many sources that are available online, such as [1]_, [2]_, [3]_ .
+
+    .. [1] Timcheck, S. W. (2017). Efficient Implementation of Reductions on GPU
+    Architectures.
+
+    .. [2] https://dournac.org/info/gpu_sum_reduction
 
-    if `size1` is None then the kernel expects 1d tensor inputs. If `size1` is not None
-    then the expected shape of input tensors is `(size0, size1)`, and the reduction
-    operation is equivalent to input.sum(axis=1). Depending on the size of the sum
-    axis, it might be worth tuning `work_group_size` for better performance.
+    .. [3] https://shreeraman-ak.medium.com/parallel-reduction-with-cuda-d0ae10c1ae2c
     """
-    if size1 is not None:
-        n_rows = size0
+    if is_1d := not size1:
+        axis = 1
+        size1 = size0
+        size0 = 1
+
+    # NB: the shape of the work group is different in each of those two cases. Summing
+    # efficiently requires to adapt to very different IO patterns depending on the sum
+    # axis, which motivates the need for different kernels with different work group
+    # sizes for each cases. As a consequence, the shape of the intermediate results
+    # in the main driver loop, and the `global_size` (total number of work items fired
+    # per call) for each kernel call, are also differents, and are variabilized with
+    # the lambda functions `get_result_shape1 and `get_global_size`.
+    if axis == 1:
+        work_group_size, kernels = _prepare_sum_reduction_2d_axis1(
+            size0, work_group_size, fused_elementwise_func, dtype, device
+        )
         sum_axis_size = size1
-    else:
-        n_rows = 1
-        sum_axis_size = size0
+        get_result_shape = lambda result_sum_axis_size: (size0, result_sum_axis_size)
+        get_global_size = (
+            lambda result_sum_axis_size: result_sum_axis_size * work_group_size * size0
+        )
 
-    # fused_unary_func is applied elementwise during the first pass on data, in the
-    # first kernel execution only.
-    work_group_size, fused_func_kernel = _make_partial_sum_reduction_2d_axis1_kernel(
-        n_rows, work_group_size, fused_unary_func, dtype, device
-    )
-    # subsequent kernel calls only sum the data.
-    _, nofunc_kernel = _make_partial_sum_reduction_2d_axis1_kernel(
-        n_rows, work_group_size, fused_unary_func=None, dtype=dtype, device=device
-    )
+    else:  # axis == 0
+        work_group_size, kernels = _prepare_sum_reduction_2d_axis0(
```

Comment:
I find this unpacking unconventional and here my proposal to improve it (2/3).
```suggestion
        (
            work_group_size,
            kernels, 
            reduction_block_size,
        ) = _prepare_sum_reduction_2d_axis0(
```